llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp

```diff
@@ -0,0 +1,256 @@
+
+bool isCast(const Value *V) {
+  int Opc = Operator::getOpcode(V);
+  return (Opc == Instruction::BitCast) || (Opc == Instruction::AddrSpaceCast);
+}
+
+using ValueSetImpl = SmallPtrSetImpl<Value *>;
+using ValueSet = SmallPtrSet<Value *, 4>;
+using ConstValueSetImpl = SmallPtrSetImpl<const Value *>;
+using ConstValueSet = SmallPtrSet<const Value *, 4>;
+
+Value *stripCasts(Value *V) {
+  if (!V->getType()->isPtrOrPtrVectorTy())
+    return V;
+  // Even though we don't look through PHI nodes, we could be called on an
+  // instruction in an unreachable block, which may be on a cycle.
+  ConstValueSet Visited;
+  Visited.insert(V);
+
+  do {
+    if (isCast(V)) {
+      V = cast<Operator>(V)->getOperand(0);
+    }
+    assert(V->getType()->isPtrOrPtrVectorTy() && "Unexpected operand type!");
+  } while (Visited.insert(V).second);
+  return V;
+}
+
+const Value *getSingleUserSkipCasts(const Value *V) {
+  while (isCast(V)) {
+    if (V->getNumUses() != 1) {
+      return nullptr;
+    }
+    V = *(V->user_begin());
+  }
+  return V;
+}
+
+void collectUsesSkipThroughCasts(Value *V, SmallPtrSetImpl<const Use *> &Uses) {
+  for (Use &U : V->uses()) {
+    Value *VV = U.getUser();
+
+    if (isCast(VV)) {
+      collectUsesSkipThroughCasts(VV, Uses);
+    } else {
+      Uses.insert(&U);
+    }
+  }
+}
+
+Value *getInvokeeIfInvokeSimdCall(const CallInst *CI) {
+  Function *F = CI->getCalledFunction();
+
+  if (F && F->getName().startswith(INVOKE_SIMD_PREF)) {
+    return CI->getArgOperand(0);
+  }
+  return nullptr;
+}
+
+void getPossibleStoredVals(Value *Addr, ValueSetImpl &Vals) {
+  ValueSet Visited;
+  AllocaInst *LocalVar = dyn_cast_or_null<AllocaInst>(stripCasts(Addr));
+
+  if (!LocalVar) {
+    llvm_unreachable("unsupported data flow pattern for invoke_simd 10");
+  }
+  SmallPtrSet<const Use *, 4> Uses;
+  collectUsesSkipThroughCasts(LocalVar, Uses);
+
+  for (const Use *U : Uses) {
+    Value *V = U->getUser();
+
+    if (auto *StI = dyn_cast<StoreInst>(V)) {
+      constexpr int StoreInstValueOperandIndex = 0;
+
+      if (U != &StI->getOperandUse(StoreInst::getPointerOperandIndex())) {
+        assert(U == &StI->getOperandUse(StoreInstValueOperandIndex));
+        // this is double indirection - not supported
+        llvm_unreachable("unsupported data flow pattern for invoke_simd 11");
+      }
+      V = stripCasts(StI->getValueOperand());
+
+      if (auto *LI = dyn_cast<LoadInst>(V)) {
+        // A value loaded from another address is stored at this address -
+        // recurse into the other address
+        getPossibleStoredVals(LI->getPointerOperand(), Vals);
+      } else {
+        Vals.insert(V);
+      }
+      continue;
+    }
+    if (const auto *CI = dyn_cast<CallInst>(V)) {
+      // only __builtin_invoke_simd is allowed, otherwise the pointer escapes
+      if (!getInvokeeIfInvokeSimdCall(CI)) {
+        llvm_unreachable("unsupported data flow pattern for invoke_simd 12");
```

Comment:
I'm just wondering why it was decided to use numerical id for errors instead of describing what is actually unexpected?
As I see there are comments that describes unsupported cases before some of these calls. Might they be used as messages?

Reply:
Ok, thank you very much for explanation.